### PR TITLE
Add JAR project compiler

### PR DESCRIPTION
**Pull request by gandreadis**
Adds a `ProjectCompiler` pipeline stage implementation for JAR projects. The compiler walks through the list of files in the JAR and, if they are class files, adds their name to the list of fully qualified class names.

### CODE REVIEW

File: modules/pipeline/java-jar-project-compiler/src/test/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompilerTest.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import org.jetbrains.spek.api.dsl.describe
+import org.jetbrains.spek.api.dsl.it
+import java.io.File
+import java.io.FileNotFoundException
+
+internal class ProjectCompilerTest : Spek({
+    val projectURI = ProjectCompilerTest::class.java.getResource("/schaapi.simple-project-1.0.0.jar")
+        ?: throw FileNotFoundException("Could not find test resources.")
+
+    describe("Java JAR project compilation") {
```
Comment (FWDekker) [inline modules/pipeline/java-jar-project-compiler/src/test/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompilerTest.kt]: I think there are some extra cases you should test:

* There are no `class` files?
* There is a JAR in the JAR?
* There is a directory that ends in `.class`?

File: modules/pipeline/java-jar-project-compiler/build.gradle [hunk 2]
```diff
@@ -0,0 +1,5 @@
+dependencies {
+    compile project(":pipeline")
+
+    compile project(":models:java-project")
+}
```
[the remark this answers is not in the document]
Comment (gandreadis) [reply modules/pipeline/java-jar-project-compiler/build.gradle]: I know that, but `java-maven-project` currently also has an empty `gradle.properties`. I added this one for consistency

File: modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
+package org.cafejojo.schaapi.pipeline.projectcompiler.javajar
+
+import org.cafejojo.schaapi.models.Project
+import org.cafejojo.schaapi.models.project.java.JavaJarProject
+import org.cafejojo.schaapi.pipeline.ProjectCompiler
+import java.io.FileInputStream
+import java.util.jar.JarInputStream
+
+/**
+ * Compiles a Java project consisting of a single JAR.
```
[the remark this answers is not in the document]
Comment (gandreadis) [reply modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt]: I agree that it doesn't really compile, but on the other hand, the whole `ProjectCompiler` name of the interface is confusing / not applicable to this module (as it does not actually compile but rather analyse).

File: modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ */
+class ProjectCompiler : ProjectCompiler {
+    /**
+     * Compiles a project.
```
[the remark this answers is not in the document]
Comment (gandreadis) [reply modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt]: Again, I fully agree, but see what `org.cafejojo.schaapi.pipeline.projectcompiler.javamaven.ProjectCompiler` does (exactly the same). Should we remove it from there, too?

File: modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+     *
+     * @param project an uncompiled project
+     * @return a compiled project
+     */
+    override fun compile(project: Project): Project {
+        if (project !is JavaJarProject) throw IllegalArgumentException("Project must be JavaJarProject.")
+
+        val classNames = mutableListOf<String>()
+        val jarInputStream = JarInputStream(FileInputStream(project.classDir))
+        while (true) {
```
[the remark this answers is not in the document]
Comment (gandreadis) [reply modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt]: I agree, I'll refactor this

File: modules/pipeline/java-jar-project-compiler/src/test/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompilerTest.kt (hunk 1 above)
Comment (gandreadis) [reply modules/pipeline/java-jar-project-compiler/src/test/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompilerTest.kt]: I'll add tests for 1/ and 3/, 2/ is out of scope for this project IMO. I already only add `class` files to the list, so any JAR inside the JAR would most likely be ignored.

Comment (gandreadis) [reply modules/pipeline/java-jar-project-compiler/src/test/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompilerTest.kt]: On a revised note, I don't know how I could produce testcase 3/ - that's not a valid directory name in a JAR (as all those names correspond to package structures, and that would not be a valid package name)

File: modules/pipeline/java-jar-project-compiler/build.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (FWDekker) [reply modules/pipeline/java-jar-project-compiler/build.gradle]: Then `java-maven-project` shouldn't have had a `gradle.properties` to begin with.

File: modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (FWDekker) [reply modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt]: Regardless, the current documentation is confusing and should explain what the class does. We can always rename the superclass later.

File: modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (FWDekker) [reply modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt]: Yes

File: modules/pipeline/java-jar-project-compiler/build.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (gandreadis) [reply modules/pipeline/java-jar-project-compiler/build.gradle]: Fixed

File: modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (gandreadis) [reply modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt]: Fixed

File: modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (gandreadis) [reply modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt]: Fixed

File: modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ */
+class ProjectCompiler : ProjectCompiler {
+    override fun compile(project: Project): Project {
+        if (project !is JavaJarProject) throw IllegalArgumentException("Project must be JavaJarProject.")
+
+        val classNames = mutableListOf<String>()
+        val jarInputStream = JarInputStream(FileInputStream(project.classDir))
+        var jarEntry = jarInputStream.nextJarEntry
+
+        while (jarEntry != null) {
```
Comment (casperboone) [inline modules/pipeline/java-jar-project-compiler/src/main/kotlin/org/cafejojo/schaapi/pipeline/projectcompiler/javajar/ProjectCompiler.kt]: Meh.... Iteration like this is soooo ugly. (but I get that there's no other option)

**Pull request merged**